common/ledger/blockledger/fileledger/factory.go

```diff
@@ -26,39 +28,54 @@ type fileLedgerFactory struct {
 	mutex              sync.Mutex
 }
 
-// GetOrCreate gets an existing ledger (if it exists) or creates it if it does not
-func (flf *fileLedgerFactory) GetOrCreate(chainID string) (blockledger.ReadWriter, error) {
-	flf.mutex.Lock()
-	defer flf.mutex.Unlock()
+// GetOrCreate gets an existing ledger (if it exists) or creates it
+// if it does not.
+func (f *fileLedgerFactory) GetOrCreate(channelID string) (blockledger.ReadWriter, error) {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
 
-	key := chainID
 	// check cache
-	ledger, ok := flf.ledgers[key]
+	ledger, ok := f.ledgers[channelID]
 	if ok {
 		return ledger, nil
 	}
 	// open fresh
-	blockStore, err := flf.blkstorageProvider.Open(key)
+	blockStore, err := f.blkstorageProvider.Open(channelID)
 	if err != nil {
 		return nil, err
 	}
 	ledger = NewFileLedger(blockStore)
-	flf.ledgers[key] = ledger
+	f.ledgers[channelID] = ledger
 	return ledger, nil
 }
 
-// ChannelIDs returns the channel IDs the factory is aware of
-func (flf *fileLedgerFactory) ChannelIDs() []string {
-	channelIDs, err := flf.blkstorageProvider.List()
+// Remove removes an existing ledger.
+func (f *fileLedgerFactory) Remove(channelID string) error {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	err := f.blkstorageProvider.Drop(channelID)
+	if err != nil {
+		return err
+	}
+
+	delete(f.ledgers, channelID)
+
+	return nil
+}
+
+// ChannelIDs returns the channel IDs the factory is aware of.
+func (f *fileLedgerFactory) ChannelIDs() []string {
+	channelIDs, err := f.blkstorageProvider.List()
```

Comment:
I recall that the ledger team reported that the blkstorageProvider may report empty ledgers in the `List`. This may happen if the orderer crashes after it created the ledger but before it appends the genesis block to it.  If that is the case, the `New()` below would be a good place to weed those out and remove them.

Unrelated to this PR - I am not sure this should panic... maybe leave the decision to an upper layer?

Reply:
Yes. The block store does not handle any failure/recovery for the consumer as it is common to both peer and orderer. Hence, it is responsibility of the consumer (in this case, orderer) to take care of failure/recovery. 

As Yoav mentioned, the `List()` cannot be trusted to return all active ledgers. To avoid misuse of `List()` in the future, we will consider to remove it altogether. A failure could happen either during channel creation or during the removal as well. In both cases, it is the responsibility of the block store consumer to perform cleanup during the reboot. In order to do that, we need to have bookkeeper at the orderer and keep track of ledger status such as `ACTIVE`, `UNDER_CONSTRUCTION`, and `UNDER_DESTRUCTION` either using the existing file repo or a new levelDB instance. During the orderer bootup, a necessary cleanup needs to be performed. We do this at peer here to delete all folders/files/data including the blockstore https://github.com/hyperledger/fabric/blob/add66376d90559d44a45bf61c6185079ac8c37ee/core/ledger/kvledger/kv_ledger_provider.go#L122

Reply:
@cendhu Do you propose we handle that in this PR or would it be fine to do as a follow-up?

Reply:
@wlahti 
```
func (flf *fileLedgerFactory) ChannelIDs() []string {
	channelIDs, err := flf.blkstorageProvider.List()
```
is an existing code and not introduced by this PR. As the goal of the current PR is to add `Remove()` function, I am okay to have a separate PR to handle the failure of both ledger creation & deletion.